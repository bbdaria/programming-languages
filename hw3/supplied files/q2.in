print "===TEST START===\n";
(* Testing the build4 function *)
val test1 = build4 (1, "a", 2, "b");
val test2 = build4 ("hello", 3.14, "world", 42.0);
val test3 = build4 (true, 1, false, 2);
val test4 = build4 (#"a", "hello", #"b", "world");

(* Testing the unzip function *)
val test5 = unzip (1::: "a"::: 2::: "b"::: NIL);
val test6 = unzip ("x"::: 1.1::: "y"::: 2.2::: NIL);
val test7 = unzip (#"a"::: "hello"::: #"b"::: "world"::: NIL);
val test8 = unzip (true:::1:::false:::2:::NIL);
val testlol = unzip ("never":::"never":::"gonna":::"gonna":::"give":::"let":::"you":::"you":::"up":::"down":::NIL)

(* Testing the zip function *)
val test9 = zip ([1, 2], ["a", "b"]);
val test10 = zip (["x", "y"], [1.1, 2.2]);
val test11 = zip ([#"a", #"b"], ["hello", "world"]);
val test12 = zip ([true, false], [1, 2]);

(* Handling exceptions in zip function tests *)
val test13 = zip ([1, 2], ["a"]) handle Empty => NIL;
val test14 = zip (["x"], [1.1, 2.2]) handle Empty => NIL;
