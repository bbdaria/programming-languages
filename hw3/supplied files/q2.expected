val it = () : unit
val test1 = 1 ::: "a" ::: 2 ::: "b" ::: NIL : (int,string) heterolist
val test2 = "hello" ::: 3.14 ::: "world" ::: 42 ::: NIL
  : (string,real) heterolist
val test3 = true ::: 1 ::: false ::: 2 ::: NIL : (bool,int) heterolist
val test4 = #"a" ::: "hello" ::: #"b" ::: "world" ::: NIL
  : (char,string) heterolist
val test5 = ([1,2],["a","b"]) : int list * string list
val test6 = (["x","y"],[1.1,2.2]) : string list * real list
val test7 = ([#"a",#"b"],["hello","world"]) : char list * string list
val test8 = ([true,false],[1,2]) : bool list * int list
val testlol =
  (["never","gonna","give","you","up"],["never","gonna","let","you","down"])
  : string list * string list
val test9 = 1 ::: "a" ::: 2 ::: "b" ::: NIL : (int,string) heterolist
val test10 = "x" ::: 1.1 ::: "y" ::: 2.2 ::: NIL : (string,real) heterolist
val test11 = #"a" ::: "hello" ::: #"b" ::: "world" ::: NIL
  : (char,string) heterolist
val test12 = true ::: 1 ::: false ::: 2 ::: NIL : (bool,int) heterolist
val test13 = NIL : (int,string) heterolist
val test14 = NIL : (string,real) heterolist
- 
